File: gaussian_relu_moments.m
function [o_mean, o_covariance] = gaussian_relu_moments(mean, covariance)
%GAUSSIAN_RELU_MOMENTS Get mean and covariance for ReLU(gaussian_vector)
%   We start by Omega from Equation (1) in the Network Moments paper
%   The expression presented in the paper for Omega is for the 2D case
%   (i.e., assuming size(mean) is [2 1] and size(covariance) is [2 2]).
%   Here we will need to generalize this for any number of dimensions n.
%   One approach is to define it for 2D and call it for every possible
%   pair of dimensions, then combine the results in a single matrix.
%   Instead, use [n n] matrices or [n 1] vectors as intermediate resutls.
%   Every element in these matrices naturally corresponds to a pair.
% 	We will be very careful with transposing and broadcasting operations.

n = numel(mean);
assert(all(size(mean) == [n 1]), 'mean must be a column vector');
assert(all(size(covariance) == [n n]), 'covariance must match mean');

%% computing Omega

% det(covariance) could be deceiving since it was written for 2D
% here, we need it for all pairs (symmetric matrix with zero diagonals)
% we can compute sigma_det as std_outer.^2 .* (1 - rho.^2) 
% where rho=covariance./std_outer; correlation coefficients (-1<=rho<=1) or
std = realsqrt(diag(covariance));
std_outer = std' .* std;
sigma_det = max(std_outer - covariance, 0) .* (std_outer + covariance);

% similarly, the term mean' * inv(covariance) * mean is true only in 2D
% here, we can compute this as follows (from the inverse of 2D matrices)
% (std_x_mean'.^2 + std_x_mean.^2 - 2*mean_outer.*covariance) ./ sigma_det
% knowing that, mean_outer .* covariance = std_x_mean .* std_x_mean' .* rho
% note here that the diagonals are undefined for this term since rho = 1
% also, the numerator and denomenator are zero-diagonal symmetric matrices
mean_outer = mean' .* mean;
std_x_mean = mean' .* std;
q = @(x, y) x' + x - 2 .* y;
quadratic = q(std_x_mean.^2, mean_outer .* covariance) ./ sigma_det;

% the first term is symmetric (i.e., term1(x1, x2) == term1(x2, x1))
sqrt_sigma_det = realsqrt(sigma_det);
term1 = (1 / (2 * pi)) .* sqrt_sigma_det .* exp(-0.5 .* quadratic);

% sigma_tilde_mean = sigma_tilde * mean / sqrt(2) (in 2D, sigma_tilde is):
% diag(sqrt(det(cov) ./ [cov(2, 2); cov(1, 1)])) * inv(cov)
% but here, the result is a non-symmetric matrix with undefined diagonals
mean_std = sqrt(1 / 2) .* mean ./ std;
sigma_tilde_mean = sqrt(1 / 2) .* std_x_mean' - mean_std' .* covariance;
sigma_tilde_mean = sigma_tilde_mean ./ sqrt_sigma_det;

% the second and third terms is an undefined-diagonal non-symmetric matrix
% because of this asymmetry, it is a bit tricky (i.e., pairs are ordered)
pdf = (1 / sqrt(2 * pi)) .* exp(-mean_std.^2);
term2 = (1 / 2) .* std_x_mean' .* pdf' .* (1 + erf(sigma_tilde_mean));

% the forth term is also non-symmetric because of cdf'
cdf = (1 / 2) .* (1 + erf(mean_std));
mean_sigma = mean_outer + covariance;
term4 = (1 / 2) .* mean_sigma .* cdf';

% now to the final result Omega
omega = term1 + term2 + term2' + term4;

%% integrating erf(a*x+b)/exp(x^2) from c to Inf

% a = rho .* irho; b = mean_std .* irho; c = -mean_std';
irho = std_outer ./ sqrt_sigma_det;  % equivalent to 1 ./ sqrt(1 - rho.^2)
a = irho .* covariance ./ std_outer;
b = irho .* mean_std;
c = repmat(-mean_std', n, 1);
b(irho == Inf) = 0;  % avoid Inf-Inf in a .* x + b (when irho is Inf)
f = @(a, b, c) integral(@(x) erf(a .* x + b) ./ exp(x.^2), c, Inf);
integration = arrayfun(f, a, b, c);  % erf_exp_integral(a, b, c);

%% correlation matrix E[ReLU(x)ReLU(x)^T]

correlation = omega + (1 / sqrt(4 * pi)) .* mean_sigma .* integration;
second_moment = diag(mean_sigma) .* cdf + diag(std_x_mean) .* pdf;
correlation(1:n + 1:end) = second_moment;  % replace the diagonal

%% mean and covariance matrix E[ReLU(x)ReLU(x)^T] - E[ReLU(x)]E[ReLU(x)]^T

o_mean = mean .* cdf + std .* pdf;
o_covariance = correlation - o_mean' .* o_mean;

end

File: gaussian_relu_moments.py
from math import pi, sqrt

from erf_exp_integral import erf_exp_integral

__all__ = [
    # main functions
    'relu_cov_mean',  # not differentiable for non-zero mean (for now)
    'relu_var_mean',
    'relu_mean',

    # auxiliary functions
    'relu_std_mean',
    'relu_cov',
    'relu_var',
    'relu_std',
    'forward_gaussian',
]


def relu_cov_mean(covariance, mean=None, num_terms=0):
    """Compute the covariance and mean of ReLU(gaussian_vector)."""
    d = lambda x: x.diagonal(0, -1, -2)  # diagonal of a matrix
    t = lambda x: x.transpose(-1, -2)  # transpose of a matrix
    o = lambda x, y: x.unsqueeze(-1) * y.unsqueeze(-2)  # vector outer product
    m = lambda x, y: x * y.unsqueeze(-2)  # multiply matrix columns by a vector

    std = d(covariance).sqrt()
    std_outer = o(std, std)
    if mean is None:
        mean = relu_mean(std)
        correlation = (covariance / std_outer).clamp(-1, 1)
        c = lambda x: (x * (-x).acos() + (1 - x.pow(2)).sqrt())
        o_correlation = (1 / (2 * pi)) * c(correlation) * std_outer
        return o_correlation - o(mean, mean), mean

    sigma_det = (std_outer - covariance).clamp(0) * (std_outer + covariance)

    mean_outer, std_x_mean = o(mean, mean), o(std, mean)
    q = lambda x, y, z: (x + t(x) - 2 * y) / z
    quadratic = q(std_x_mean.pow(2), mean_outer * covariance, sigma_det)

    sqrt_sigma_det = sigma_det.sqrt()
    term1 = (1 / (2 * pi)) * sqrt_sigma_det * (-0.5 * quadratic).exp()

    mean_std = sqrt(1 / 2) * (mean / std)
    n = lambda x, y, z: (sqrt(1 / 2) * t(x) - y) / z
    sigma_tilde_mean = n(std_x_mean, m(covariance, mean_std), sqrt_sigma_det)

    pdf = (1 / sqrt(2 * pi)) * (-mean_std.pow(2)).exp()
    term2 = (1 / 2) * m(t(std_x_mean), pdf) * (1 + sigma_tilde_mean.erf())

    cdf = (1 / 2) * (1 + mean_std.erf())
    mean_sigma = mean_outer + covariance
    term4 = (1 / 2) * m(mean_sigma, cdf)

    omega = term1 + term2 + t(term2) + term4

    irho = std_outer / sqrt_sigma_det
    a = irho * covariance / std_outer
    b = irho * mean_std.unsqueeze(-1)
    c = -mean_std.unsqueeze(-2)
    b[(irho == float('inf')).expand_as(b)] = 0
    integration = erf_exp_integral(a, b, c, num_terms - 1)

    correlation = omega + (1 / sqrt(4 * pi)) * mean_sigma * integration
    second_moment = d(mean_sigma) * cdf + d(std_x_mean) * pdf
    d(correlation).copy_(second_moment)

    o_mean = mean * cdf + std * pdf
    o_covariance = correlation - o(o_mean, o_mean)
    return o_covariance, o_mean


def relu_var_mean(std, mean=None):
    """Compute the variance and mean of ReLU(gaussian_vector)."""
    if mean is None:
        o_mean = relu_mean(std)
        return o_mean * std * ((pi - 1) / sqrt(2 * pi)), o_mean
    d_std_x_mean = std * mean
    d_mean_sigma = mean.pow(2) + std.pow(2)
    mean_std = sqrt(1 / 2) * mean / std
    pdf = (1 / sqrt(2 * pi)) * (-mean_std.pow(2)).exp()
    cdf = (1 / 2) * (1 + mean_std.erf())
    o_mean = mean * cdf + std * pdf
    second_moment = d_mean_sigma * cdf + d_std_x_mean * pdf
    return second_moment - o_mean.pow(2), o_mean


def relu_mean(std, mean=None):
    """Compute the mean of ReLU(gaussian_vector)."""
    if mean is None:
        return std * (1 / sqrt(2 * pi))
    mean_std = sqrt(1 / 2) * mean / std
    pdf = (1 / sqrt(2 * pi)) * (-mean_std.pow(2)).exp()
    cdf = (1 / 2) * (1 + mean_std.erf())
    return mean * cdf + std * pdf


def relu_std_mean(std, mean=None):
    """Compute the standard deviation and mean of ReLU(gaussian_vector)."""
    variance, mean = relu_var_mean(std, mean)
    return variance.sqrt(), mean


def relu_cov(covariance, mean=None, num_terms=0):
    """Compute the covariance of ReLU(gaussian_vector)."""
    return relu_cov_mean(covariance, mean, num_terms)[0]


def relu_var(std, mean=None):
    """Compute the variance of ReLU(gaussian_vector)."""
    return relu_var_mean(std, mean)[0]


def relu_std(std, mean=None):
    """Compute the standard deviation of ReLU(gaussian_vector)."""
    return relu_var_mean(std, mean)[0].sqrt()


def forward_gaussian(model, covariance, mean, terms=5):
    """Compute variance and mean of Affine-ReLU-Affine for gaussian input."""
    a, _, b = model  # assumes Sequential(Linear, ReLU, Linear)
    covariance, mean = a.weight @ covariance @ a.weight.t(), a(mean)
    if terms >= 0:
        covariance, mean = relu_cov_mean(covariance, mean, terms)
    else:  # assume zero mean relu input to compute off-diagonals
        std = covariance.diagonal(0, -1, -2).sqrt()
        variance, mean = relu_var_mean(std, mean)
        covariance = relu_cov(covariance)
        covariance.diagonal(0, -1, -2).copy_(variance)
    variance, mean = ((b.weight @ covariance) * b.weight).sum(-1), b(mean)
    return variance, mean
